src/auth/src/hash.rs

```diff
@@ -0,0 +1,230 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+// Clippy misreads some doc comments as HTML tags, so we disable the lint
+#![allow(rustdoc::invalid_html_tags)]
+
+use std::fmt::Display;
+use std::num::NonZeroU32;
+
+use base64::prelude::*;
+use ring::digest::SHA256_OUTPUT_LEN;
+use ring::hmac::{self, Key, HMAC_SHA256};
+use ring::pbkdf2::{self, PBKDF2_HMAC_SHA256 as SHA256};
+use ring::rand::SecureRandom;
+
+/// The default iteration count as suggested by
+/// https://cheatsheetseries.owasp.org/cheatsheets/Password_Storage_Cheat_Sheet.html
+const DEFAULT_ITERATIONS: NonZeroU32 = NonZeroU32::new(600_000).unwrap();
+
+/// The default salt size, which isn't currently configurable.
+const DEFAULT_SALT_SIZE: usize = 32;
+
+/// The options for hashing a password
+pub struct HashOpts {
+    /// The number of iterations to use for PBKDF2
+    pub iterations: NonZeroU32,
+    /// The salt to use for PBKDF2. It is up to the caller to
+    /// ensure that however the salt is generated, it is cryptographically
+    /// secure.
+    pub salt: [u8; DEFAULT_SALT_SIZE],
+}
+
+pub struct PasswordHash {
+    /// The salt used for hashing
+    pub salt: [u8; DEFAULT_SALT_SIZE],
+    /// The number of iterations used for hashing
+    pub iterations: NonZeroU32,
+    /// The hash of the password.
+    /// This is the result of PBKDF2 with SHA256
+    pub hash: [u8; SHA256_OUTPUT_LEN],
+}
+
+#[derive(Debug)]
+pub enum VerifyError {
+    MalformedHash,
+    InvalidPassword,
+}
+
+/// Hashes a password using PBKDF2 with SHA256
+/// and a random salt.
+pub fn hash_password(password: &String) -> PasswordHash {
```

Comment:
nit: should this also be `Password`

```suggestion
pub fn hash_password(password: &Password) -> PasswordHash {
```